src/google-maps/map-marker-clusterer/map-marker-clusterer.ts

```diff
@@ -0,0 +1,603 @@
+        this.markerClusterer.setImagePath(imagePath);
+      }
+    });
+  }
+
+  private _watchForImageSizesChanges() {
+    this._imageSizes.pipe(takeUntil(this._destroy)).subscribe(imageSizes => {
+      if (this.markerClusterer && imageSizes) {
+        this._assertInitialized();
+        this.markerClusterer.setImageSizes(imageSizes);
+      }
+    });
+  }
+
+  private _watchForMaxZoomChanges() {
+    this._maxZoom.pipe(takeUntil(this._destroy)).subscribe(maxZoom => {
+      if (this.markerClusterer && maxZoom !== undefined) {
+        this._assertInitialized();
+        this.markerClusterer.setMaxZoom(maxZoom);
+      }
+    });
+  }
+
+  private _watchForMinimumClusterSizeChanges() {
+    this._minimumClusterSize.pipe(takeUntil(this._destroy)).subscribe(minimumClusterSize => {
+      if (this.markerClusterer && minimumClusterSize !== undefined) {
+        this._assertInitialized();
+        this.markerClusterer.setMinimumClusterSize(minimumClusterSize);
+      }
+    });
+  }
+
+  private _watchForStylesChanges() {
+    this._styles.pipe(takeUntil(this._destroy)).subscribe(styles => {
+      if (this.markerClusterer && styles) {
+        this._assertInitialized();
+        this.markerClusterer.setStyles(styles);
+      }
+    });
+  }
+
+  private _watchForTitleChanges() {
+    this._title.pipe(takeUntil(this._destroy)).subscribe(title => {
+      if (this.markerClusterer && title !== undefined) {
+        this._assertInitialized();
+        this.markerClusterer.setTitle(title);
+      }
+    });
+  }
+
+  private _watchForZIndexChanges() {
+    this._zIndex.pipe(takeUntil(this._destroy)).subscribe(zIndex => {
+      if (this.markerClusterer && zIndex !== undefined) {
+        this._assertInitialized();
+        this.markerClusterer.setZIndex(zIndex);
+      }
+    });
+  }
+
+  private _watchForZoomOnClickChanges() {
+    this._zoomOnClick.pipe(takeUntil(this._destroy)).subscribe(zoomOnClick => {
+      if (this.markerClusterer && zoomOnClick !== undefined) {
+        this._assertInitialized();
+        this.markerClusterer.setZoomOnClick(zoomOnClick);
+      }
+    });
+  }
+
+  private _watchForMarkerChanges() {
+    this._assertInitialized();
+    const initialMarkers: google.maps.Marker[] = [];
+    for (const marker of this._getInternalMarkers(this._markers.toArray())) {
+      this._currentMarkers.add(marker);
+      initialMarkers.push(marker);
+    }
+    this.markerClusterer.addMarkers(initialMarkers);
+
+    this._markers.changes.pipe(takeUntil(this._destroy)).subscribe(markerComponents => {
+      this._assertInitialized();
+      const newMarkers = new Set<google.maps.Marker>(markerComponents
+        .filter((markerComponent: MapMarker) => !!markerComponent.marker)
+        .map((markerComponent: MapMarker) => markerComponent.marker));
+      const markersToAdd: google.maps.Marker[] = [];
+      const markersToRemove: google.maps.Marker[] = [];
+      for (const marker of Array.from(newMarkers)) {
```

Comment:
It seems like we're making a lot of conversions between sets and maps, as well as creating new arrays. Since this component is supposed to be used when a lot of markers are on the map, maybe we should optimize it a bit?

Reply:
The idea behind the complexity here is twofold: One is to avoid the most costly operation, the repainting, until the very end when all of the adding and removing has been figured out, and two is to avoid any quadratic time operations. There are a lot of conversions between arrays and sets, but the linear time complexity involved with those operations is paltry compared to if we had to search or sort an array every time we needed to add or remove a marker. 

So, I think this is actually optimal, are there any additional improvements you see?

Reply:
Off the top of my head, I think that we could get rid of at least two new arrays by replacing the `filter` and `map` above with a single `for` loop with an `if` statement.

Reply:
Done.